### ISSUE TITLE
Possible data loss during deployment

### ISSUE THREAD

**Issue by osrf-migration** [osrf/cloudsim-grant#14]
**Original report ([archived issue](https://osrf-migration.github.io/osrf-archived-gh-pages/#!/osrf/cloudsim-grant/issues/14)) by Hugo Boyer (Bitbucket: [hugomatic](https://bitbucket.org/%7B3d87874e-7099-4981-b04f-57f0272faa7d%7D/), GitHub: [hugomatic](https://github.com/hugomatic)).**

----------------------------------------
The auth server on the production environment was missing user data.

This problem could happen when the deployment is interrupted. 
It was also noticed that deployment of the portal was taking a long time (5 minutes)

These problems may be related to the code that reads the database data (using too many simultaneous asynchronous calls in parallel)

**Comment by osrf-migration** [comment]
**Original comment by Hugo Boyer (Bitbucket: [hugomatic](https://bitbucket.org/%7B3d87874e-7099-4981-b04f-57f0272faa7d%7D/), GitHub: [hugomatic](https://github.com/hugomatic)).**

----------------------------------------

* set assignee_account_id to "557058:869c02bb-b246-4cdc-9db3-0064e8add185"
* set assignee to "hugomatic (Bitbucket: [hugomatic](https://bitbucket.org/hugomatic), GitHub: [hugomatic](https://github.com/hugomatic))"

I've been working on serializing all the database access code (in the actions branch).
There are deployment issues on certain systems, I think I've hit the following bug:

https://github.com/NodeRedis/node-redis-parser/issues/7

**Comment by osrf-migration** [comment]
**Original comment by Hugo Boyer (Bitbucket: [hugomatic](https://bitbucket.org/%7B3d87874e-7099-4981-b04f-57f0272faa7d%7D/), GitHub: [hugomatic](https://github.com/hugomatic)).**

----------------------------------------
New findings. Since the data corruption has only been observed in our user data, and there's an outstanding bug reported on the "Cloudsim status spreadsheet" about missing user data, it looks like the problem is in the auth server, not the data reading code.

The auth server in dev does this:
```
#!javascript

if (process.env.CLOUDSIM_ADMIN)
  adminUser = process.env.CLOUDSIM_ADMIN;
csgrant.init(adminUser, {'root': {}, 'group':{} }, dbName,
  process.env.CLOUDSIM_AUTH_DB, ()=>{
    console.log( dbName + ' redis database loaded')
  });

...

if (process.env.CLOUDSIM_ADMIN)
  adminUser = process.env.CLOUDSIM_ADMIN;
csgrant.init(adminUser, {'root': {}, 'group':{} }, dbName,
  process.env.CLOUDSIM_AUTH_DB, ()=>{
    console.log( dbName + ' redis database loaded')
  });


```



The auth server in production is newer, and does this:
```
#!javascript


csgrant.init(resources,
  dbName,
  process.env.CLOUDSIM_AUTH_DB,
  httpServer,
  ()=>{
    console.log( dbName + ' redis database loaded')
    httpServer.listen(port)
  })


```
I believe that starting the server in parallel with the server (old dev code) can cause the problem we've seen. 

I believe we lost user data when I upgraded the production database machine from a t2.micro. Changing the database ip on Elastic Beanstalk triggered a deployment (which is rare for the auth server).

I'm in the process of replicating the data and trying to reproduce the problem on my local machine.

I think that the fix I was working on was not the correct one, but has value since it resumes previous work (to enable batch grant/revoke) and provides many new unit tests, including a test server.

Since the current "default" of the auth server already includes better data loading, expect a PR on auth (but no new issue).

**Comment by osrf-migration** [comment]
**Original comment by Hugo Boyer (Bitbucket: [hugomatic](https://bitbucket.org/%7B3d87874e-7099-4981-b04f-57f0272faa7d%7D/), GitHub: [hugomatic](https://github.com/hugomatic)).**

----------------------------------------

* changed state from "new" to "resolved"

Reproduced and fixed in:
https://osrf-migration.github.io/osrf-archived-gh-pages/#!/osrf/cloudsim-auth/pull-requests/21

The work on the actions branch will be deployed separately.
